File: ai-analytics-service/src/schemas/scoring_bands.py
"""Score bands, read from the manifest both runtimes share.

`status_for_score` used to exist twice here and three more times in Core, with
the cross-runtime corpus as the only thing that would notice a copy drifting.
All of them now come from `contracts/scoring-bands.json`, which Core reads
through `src/lib/scoring-bands.ts`.

The status alias lives here rather than in `mcp_types` so that the module every
runtime boundary depends on has no dependencies of its own; `mcp_types`
re-exports it, and its importers are unaffected.
"""

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, List, Literal

DimensionStatus = Literal["green", "yellow", "red"]

_STATUS_ORDER: List[DimensionStatus] = ["green", "yellow", "red"]
_LOWEST_SCORE = 0
_HIGHEST_SCORE = 100


@dataclass(frozen=True)
class ScoreBand:
    status: DimensionStatus
    min: int
    max: int


def load_scoring_bands(data: Any) -> List[ScoreBand]:
    if not isinstance(data, dict):
        raise ValueError("Scoring bands manifest must be an object")
    raw_bands = data.get("bands")
    if not isinstance(raw_bands, list) or len(raw_bands) != len(_STATUS_ORDER):
        raise ValueError(
            f"Scoring bands manifest must define {len(_STATUS_ORDER)} bands",
        )

    bands: List[ScoreBand] = []
    for index, raw_band in enumerate(raw_bands):
        if not isinstance(raw_band, dict):
            raise ValueError(f"Scoring band {index} must be an object")
        expected_status = _STATUS_ORDER[index]
        if raw_band.get("status") != expected_status:
            raise ValueError(
                f"Scoring band {index} must have status '{expected_status}'",
            )
        minimum = raw_band.get("min")
        maximum = raw_band.get("max")
        # bool is an int in Python, and a band bound of True would otherwise
        # pass every check below and quietly behave as 1.
        if (
            not isinstance(minimum, int)
            or not isinstance(maximum, int)
            or isinstance(minimum, bool)
            or isinstance(maximum, bool)
        ):
            raise ValueError(
                f"Scoring band '{expected_status}' needs integer min and max",
            )
        if minimum > maximum:
            raise ValueError(
                f"Scoring band '{expected_status}' has min above max",
            )
        bands.append(ScoreBand(status=expected_status, min=minimum, max=maximum))

    # Descending order with no gap and no overlap, so every score has exactly
    # one colour whichever runtime asks.
    if bands[0].max != _HIGHEST_SCORE:
        raise ValueError(f"Scoring bands must reach {_HIGHEST_SCORE}")
    if bands[-1].min != _LOWEST_SCORE:
        raise ValueError(f"Scoring bands must start at {_LOWEST_SCORE}")
    for previous, current in zip(bands, bands[1:]):
        if current.max + 1 != previous.min:
            raise ValueError(
                f"Scoring bands '{previous.status}' and '{current.status}' "
                "must be adjacent",
            )

    return bands


def _load_bands() -> List[ScoreBand]:
    # Walk up from src/schemas to the repository root, then down to contracts,
    # the same way contract_registry.py finds capabilities.json.
    project_root = Path(__file__).resolve().parent.parent.parent.parent
    bands_path = project_root / "contracts" / "scoring-bands.json"

    with bands_path.open("r", encoding="utf-8") as bands_file:
        data = json.load(bands_file)

    return load_scoring_bands(data)


SCORING_BANDS: List[ScoreBand] = _load_bands()


def status_for_score(score: float) -> DimensionStatus:
    """Map a 0-100 score onto its band.

    A score outside that range cannot come from an aggregate, so it takes the
    nearest band rather than inventing a fourth status.
    """
    for band in SCORING_BANDS:
        if score >= band.min:
            return band.status
    return SCORING_BANDS[-1].status

File: src/lib/scoring-bands.ts
import scoringBandsManifest from '../../contracts/scoring-bands.json';
import type { WellbeingStatus } from './shalomut-source';

/**
 * The one place a 0-100 score becomes a colour.
 *
 * The bands used to be three literal copies in Core and two more in Python, and
 * the only thing standing between them and a silent disagreement was the
 * cross-runtime corpus. They now come from `contracts/scoring-bands.json`,
 * which `ai-analytics-service/src/schemas/scoring_bands.py` reads as well, so
 * a methodology change after the pilot is one edit rather than six.
 *
 * This deliberately stays deployment-wide rather than round-scoped: Python
 * validates that a payload's status matches its score, so per-round bands
 * would have to travel inside the payload, which is new semantics for
 * contracts `1.0`-`6.0` and would need a version of its own.
 */
export interface ScoreBand {
  status: WellbeingStatus;
  min: number;
  max: number;
}

const STATUS_ORDER: WellbeingStatus[] = ['green', 'yellow', 'red'];
const LOWEST_SCORE = 0;
const HIGHEST_SCORE = 100;

function isInteger(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value);
}

export function loadScoringBands(manifest: unknown): ScoreBand[] {
  if (!manifest || typeof manifest !== 'object' || Array.isArray(manifest)) {
    throw new Error('Scoring bands manifest must be an object');
  }
  const rawBands = (manifest as { bands?: unknown }).bands;
  if (!Array.isArray(rawBands) || rawBands.length !== STATUS_ORDER.length) {
    throw new Error(
      `Scoring bands manifest must define ${STATUS_ORDER.length} bands`,
    );
  }

  const bands: ScoreBand[] = rawBands.map((rawBand, index) => {
    if (!rawBand || typeof rawBand !== 'object' || Array.isArray(rawBand)) {
      throw new Error(`Scoring band ${index} must be an object`);
    }
    const band = rawBand as Record<string, unknown>;
    const expectedStatus = STATUS_ORDER[index];
    if (band.status !== expectedStatus) {
      throw new Error(
        `Scoring band ${index} must have status '${expectedStatus}'`,
      );
    }
    if (!isInteger(band.min) || !isInteger(band.max)) {
      throw new Error(
        `Scoring band '${expectedStatus}' needs integer min and max`,
      );
    }
    if (band.min > band.max) {
      throw new Error(`Scoring band '${expectedStatus}' has min above max`);
    }
    return { status: expectedStatus, min: band.min, max: band.max };
  });

  // Descending order with no gap and no overlap. A gap would leave a score
  // without a colour, and an overlap would make the answer depend on which
  // copy of the lookup ran.
  if (bands[0].max !== HIGHEST_SCORE) {
    throw new Error(`Scoring bands must reach ${HIGHEST_SCORE}`);
  }
  if (bands[bands.length - 1].min !== LOWEST_SCORE) {
    throw new Error(`Scoring bands must start at ${LOWEST_SCORE}`);
  }
  for (let index = 1; index < bands.length; index++) {
    if (bands[index].max + 1 !== bands[index - 1].min) {
      throw new Error(
        `Scoring bands '${bands[index - 1].status}' and ` +
          `'${bands[index].status}' must be adjacent`,
      );
    }
  }

  return bands;
}

export const SCORING_BANDS: ScoreBand[] = loadScoringBands(
  scoringBandsManifest,
);

/**
 * Map a 0-100 score onto its band. A score outside that range cannot come from
 * an aggregate, so it takes the nearest band rather than inventing a fourth
 * status.
 */
export function statusForScore(score: number): WellbeingStatus {
  const band = SCORING_BANDS.find((candidate) => score >= candidate.min);
  return band ? band.status : SCORING_BANDS[SCORING_BANDS.length - 1].status;
}
